src/include/miopen/graphapi/opgraph.h

```diff
@@ -0,0 +1,223 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/graphapi/tensor.hpp>
+
+#include <algorithm>
+#include <unordered_map>
+#include <vector>
+
+#include <cassert>
+
+namespace miopen {
+namespace graphapi {
+
+namespace internal {
+template <typename C>
+bool contains(const C& container, const typename C::value_type& val) noexcept
+{
+    return std::find(container.cbegin(), container.cend(), val) != container.cend();
+}
+} // end namespace internal
+
+class OpGraphBuilder;
+class OpGraph;
+
+class OpNode
+{
+public:
+    using Edge = std::pair<OpNode*, Tensor*>;
+    virtual ~OpNode();
+
+private:
+    std::vector<Edge> mInEdges;
+    std::vector<Edge> mOutEdges;
+
+    friend class OpGraphBuilder;
+    friend class OpGraph;
+
+protected:
+    virtual const std::string& signName() const = 0;
+
+    virtual std::vector<Tensor*> getInTensors() const = 0;
+
+    virtual std::vector<Tensor*> getOutTensors() const = 0;
+
+    const auto& iterateInEdges() const { return mInEdges; }
+
+    const auto& iterateOutEdges() const { return mOutEdges; }
```

Comment:
It sounds like a `visitor`, but actually it's a `getter`.

Reply:
Can rename to `get`.